Update GeometryPath computations to use Point methods

aseth1 opened the pull request:
Computations by `GeometryPath` to determine the location of and speed between points are delegated to `Point`, which is now performing these calculations and reduce duplication. This greatly simplifies `GeometryPath`, but more importantly allows us to eliminate a major source of property mangling during a simulation: `GeometryPath` no longer updates a `MovingPathPoint`'s `location` property during a simulation. We are a lot closer towards enforcing that properties must be `const` during a simulation (`GeometryPath` being the primary violator). I agree with arguments that `MovingPathPoint` should not even have a `location` property, which it inherits from its `Station` super class. `PathPoint` then, should not be a `Station` either. I will leave that for another PR. 

I also made  `PathPoint` (and its children) conform to the Component API and eliminated custom *connect* like methods, which required a confusing call sequence (e.g. called when adding a Point). From the API perspective there was no need or use of the `PathPoint`'s reference (pointer)  back to a `GeometryPath` and all the `connectToModelAndPath` and `connectToModelAndBody` were removed. We are one step closer to eliminating `PathPoint` altogether and having `GeometryPath` being composed by `Points`. 

Nonetheless, the updated PathPoints are included in `testComponents` with some added handling for `Connector<Coordinate>`.

I detected no change in test performances (timings) due to these changes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/990)
<!-- Reviewable:end -->

chrisdembia commented:
*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH7T6AYUkZrEGzafIq:-KJH7T6AYUkZrEGzafIr:867476730) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
> ```cpp
> void PathWrap::extendConnectToModel(Model& model)
> {
>     _path = dynamic_cast<const GeometryPath*>(&getParent());
> ```

Missing a call to `Super::extendConnectToModel()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/ConditionalPathPoint.h, line 76 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH8Nl2LKzEJiXschfH:-KJH8Nl3FppiSEeZC9sR:512643733) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/ConditionalPathPoint.h#L76)):*
> ```objc
>     void constructConnectors() override;
>     void updateFromXMLNode(SimTK::Xml::Element& aNode,
>                            int versionNumber = -1) override;
> ```

As far as I can tell, we do not have a convention of having a default value for `versionNumber`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 337 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH9axFlkg7AKI3ycyr:-KJH9axFlkg7AKI3ycys:94168765) ([raw file](https://github.com/opensim-org/opensim-core/blob/a0a0d99ab746cf9eae42d06509d74c56568af8e3/OpenSim/Simulation/Model/MovingPathPoint.cpp#L337)):*
> ```cpp
>         msg += "must depend on same the coordinate. Condition: ";
>         msg += "x_coordinate == y_coordinate == z_coordinate  FAILED.";
>         throw Exception(msg, __FILE__, __LINE__);       
> ```

Could you explain why it is okay to remove this check?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 98 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHZJdoxReDGg2Ww1sq:-KJHZJdoxReDGg2Ww1sr:359729847) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L98)):*
> ```objc
>     // Get the local location of the MovingPathPoint in its Frame
>     SimTK::Vec3 getLocation(const SimTK::State& s) const;
>     // Get the local velocity of the MovingPathPoint in its Frame
> ```

I think this should be a doxygen comment. And it should also clarify if whether the derivative is taken in the PhysicalFrame.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 98 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHZJdoxReDGg2Ww1sq:-KJHZaqayjSGydm1Jane:-1339279731) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L98)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> I think this should be a doxygen comment. And it should also clarify if whether the derivative is taken in the PhysicalFrame.
</details>
Also, the comment for `getLocation()` should be a doxygen comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 61 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHZjeuF587tHdBJwso:-KJHZjeuF587tHdBJwsp:-1324464029) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L61)):*
> ```objc
> //==============================================================================
>     OpenSim_DECLARE_OPTIONAL_PROPERTY(x_location, Function,
>         "Function defining the x component of the point's location.");
> ```

Clarify that the location is expressed in the PhysicalFrame of the point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 61 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHZjeuF587tHdBJwso:-KJHZoDnV8oNKmkgQbOu:-1704313357) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L61)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Clarify that the location is expressed in the PhysicalFrame of the point.
</details>
er, the PhysicalFrame to which the point is connected.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 359 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH_3l4-WdepA4WVMli:-KJH_3l4-WdepA4WVMlj:1604271775) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L359)):*
> ```cpp
>     //TODO: Enable Exception or Implement the method and add accompanying test.
>     //OPENSIM_THROW(Exception, "MovingPathPoint::calcAccelerationInGround not implemented.");
>     return Vec3(SimTK::NaN);
> ```

Feel free to ignore, but the exception is preferable to the NaN. I assume you ran into some issue when trying to use the exception. Perhaps it would be good to document that issue with your TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 353 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH_HSfo8KGxfMg2Cs-:-KJH_HSgSSkY_zWWwgSt:1599241507) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L353)):*
> ```cpp
>     // linear (vF = V_GF[1]) and angular (omegaF = V_GF[0]) velocity, such that
>     // velocity of the station: v = vF + omegaF x r
>     return V_GF[1] + V_GF[0] % r + v;
> ```

I still dislike all of this math in classes like this. It just seems like we should be using methods in Simbody or on Station to do these calculations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 337 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH9axFlkg7AKI3ycyr:-KJHa43yDYmyz3XYekGW:343838915) ([raw file](https://github.com/opensim-org/opensim-core/blob/a0a0d99ab746cf9eae42d06509d74c56568af8e3/OpenSim/Simulation/Model/MovingPathPoint.cpp#L337)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Could you explain why it is okay to remove this check?
</details>
Perhaps removing this check requires mention in the CHANGELOG.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/Point.h, line 111 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHbtk83p_tRa_2WxEU:-KJHbtk83p_tRa_2WxEV:273072193) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/Point.h#L111)):*
> ```objc
>         const SimTK::Vec3& location) const;
>     /** Calculate the speed of motion between this Point and some other Point.
>     A positive speed is growing apart and negative speed is coming closer.
> ```

It would be clearer to me to say "the rate of change of the distance," if that is what it is. I'm fine with the name of the method being `calcSpeedBetween()` though. I'm not sure I understand the implementation of this method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/Point.cpp, line 134 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHdA9oqzmtphXr9MKN:-KJHdA9oqzmtphXr9MKO:-47187412) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/Point.cpp#L134)):*
> ```cpp
>     auto r = getLocationInGround(s) - o.getLocationInGround(s);
>     double d = r.norm();
>     auto v = getVelocityInGround(s) - o.getVelocityInGround(s);
> ```

Nitpicky: you could use `const auto r` and `const auto v`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
My biggest issue is that `MovingPathPoint` is a `Station`. It is *not* a Station. A Station is a point fixed to a Body.

<details><summary><i>Previously, aseth1 (Ajay Seth) wrote…</i></summary>
> #### Update GeometryPath computations to use Point methods
> 
> Computations by `GeometryPath` to determine the location of and speed between points are delegated to `Point`, which is now performing these calculations and reduce duplication. This greatly simplifies `GeometryPath`, but more importantly allows us to eliminate a major source of property mangling during a simulation: `GeometryPath` no longer updates a `MovingPathPoint`'s `location` property during a simulation. We are a lot closer towards enforcing that properties must be `const` during a simulation (`GeometryPath` being the primary violator). I agree with arguments that `MovingPathPoint` should not even have a `location` property, which it inherits from its `Station` super class. `PathPoint` then, should not be a `Station` either. I will leave that for another PR. 
> 
> 
> 
> I also made  `PathPoint` (and its children) conform to the Component API and eliminated custom *connect* like methods, which required a confusing call sequence (e.g. called when adding a Point). From the API perspective there was no need or use of the `PathPoint`'s reference (pointer)  back to a `GeometryPath` and all the `connectToModelAndPath` and `connectToModelAndBody` were removed. We are one step closer to eliminating `PathPoint` altogether and having `GeometryPath` being composed by `Points`. 
> 
> 
> 
> Nonetheless, the updated PathPoints are included in `testComponents` with some added handling for `Connector<Coordinate>`.
> 
> 
> 
> I detected no change in test performances (timings) due to these changes.
> 
> 
> 
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJHdScr2CoI8RalaNZ2:-1809171097)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
I realize that this PR did not introduce the fact that `MovingPathPoint` is a `Station`. Nonetheless, I'm stopping my review for now until we can discuss that.

<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> My biggest issue is that `MovingPathPoint` is a `Station`. It is *not* a Station. A Station is a point fixed to a Body.
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJHdmVR8DtLt_INFxNc:1642502505)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
Reviewed 6 of 17 files at r1.
Review status: 6 of 17 files reviewed at latest revision, 9 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
Ah I see you mentioned this in your PR description. Could you elaborate on any possible consequences of this, and temporarily document this in doxygen as a bug in MovingPathPoint? A component that wants to connect to a Station could actually be connected to a MovingPathPoint, causing unphysical behavior.

<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> I realize that this PR did not introduce the fact that `MovingPathPoint` is a `Station`. Nonetheless, I'm stopping my review for now until we can discuss that.
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJHeCZqKzeI9-Dzebvi:1578285339)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH7T6AYUkZrEGzafIq:-KJHfgektaVbrXdReulG:-1875309029) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Missing a call to `Super::extendConnectToModel()`?
</details>
Consider throwing an exception if the parent is not a GeometryPath.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHfvxPD3aI2agrTK7F:-KJHfvxPD3aI2agrTK7G:-1868760122) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
> ```cpp
> void PathWrap::extendConnectToModel(Model& model)
> {
>     _path = dynamic_cast<const GeometryPath*>(&getParent());
> ```

Missing a call to Super::extendConnectToModel()?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
It would be good to know your plan for addressing the issue that MovingPathPoint is a Station. One hack to help prevent mis-use for now is for MovingPathPoint to override the `location` property and set it to NaN. Not sure if that would mess up other things...

<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Ah I see you mentioned this in your PR description. Could you elaborate on any possible consequences of this, and temporarily document this in doxygen as a bug in MovingPathPoint? A component that wants to connect to a Station could actually be connected to a MovingPathPoint, causing unphysical behavior.
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJHh1XO0g3mu1wp6Csj:-1379753681)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 336 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHha4LPW7l_wsUwrhW:-KJHha4LPW7l_wsUwrhX:1135854336) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L336)):*
> ```cpp
> SimTK::Vec3 MovingPathPoint::calcLocationInGround(const SimTK::State& s) const
> {
>     return getParentFrame().getTransformInGround(s)*getLocation(s);
> ```

This tells me (as does similar code elsewhere) that there should be a `Frame::expressVectorInGround()`. I don't remember if that was discussed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 112 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHj6BKGf0TjZINvZI6:-KJHj6BLRfKXro6eE0am:-919867296) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L112)):*
> ```objc
>     SimTK::Vec3 calcLocationInGround(const SimTK::State& state) const override;
>     SimTK::Vec3 calcVelocityInGround(const SimTK::State& state) const override;
>     SimTK::Vec3 calcAccelerationInGround(const SimTK::State& state) const override;
> ```

These methods should be `final` in Station, once we remove `Station` as `MovingPathPoint`'s base class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/ConditionalPathPoint.cpp, line 79 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHjbZpgUh_4Ub6OmV4:-KJHjbZqVCBjVJ8gE4VV:139744598) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/ConditionalPathPoint.cpp#L79)):*
> ```cpp
>         if (coord != node.element_end())
>             coord->getValueAs<std::string>(coord_name);
>         XMLDocument::addConnector(node, "Connector_Coordinate_", "coordinate", coord_name);
> ```

The connectee name specified here would ideally be the path to the coordinate, not just its name. Don't worry about it for this PR though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
Reviewed 11 of 17 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

aseth1 commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 359 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH_3l4-WdepA4WVMli:-KJI5zfoK1QCQaFqYCHD:1782883518) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L359)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Feel free to ignore, but the exception is preferable to the NaN. I assume you ran into some issue when trying to use the exception. Perhaps it would be good to document that issue with your TODO.
</details>
Yes, the Exception causes testComponents' automatic verification of Outputs to fail. My preference is to simply implement it but need to also come up with test case(s) so holding off for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

aseth1 commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 353 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH_HSfo8KGxfMg2Cs-:-KJI76RscR8Z9LyKHHTA:-751393493) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L353)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> I still dislike all of this math in classes like this. It just seems like we should be using methods in Simbody or on Station to do these calculations.
</details>
It looks like Sherm intended to have similar methods in Simbody (e.g. `MobilizedBody::calcBodyMovingPointVelocityInBody()` but it is not implemented yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

aseth1 commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 336 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHha4LPW7l_wsUwrhW:-KJIcKuxOnqDVjeKGZbo:155462457) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L336)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> This tells me (as does similar code elsewhere) that there should be a `Frame::expressVectorInGround()`. I don't remember if that was discussed.
</details>
I agree that Frame::findLocationInGround and expressVectorInGround would be generally useful. Can that be a separate PR to augment Frame?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 336 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHha4LPW7l_wsUwrhW:-KJIdiAfjQOX48UaAlVd:160394146) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.cpp#L336)):*
<details><summary><i>Previously, aseth1 (Ajay Seth) wrote…</i></summary>
> I agree that Frame::findLocationInGround and expressVectorInGround would be generally useful. Can that be a separate PR to augment Frame?
</details>
Yes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

aseth1 commented:
We should not have Path specific Points of any kind. They should be considered "path" Points by virtue of being subcomponents of a GeometryPath. Given the current pace of thinning out PathPoints and updates to GeometryPath this could be achieved before releasing 4.0.

I like you idea of overriding any user setting for location and making the location NaN until we have a proper fix in place.

<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> It would be good to know your plan for addressing the issue that MovingPathPoint is a Station. One hack to help prevent mis-use for now is for MovingPathPoint to override the `location` property and set it to NaN. Not sure if that would mess up other things...
</details>


---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[OpenSim/Simulation/Model/ConditionalPathPoint.cpp, line 79 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHjbZpgUh_4Ub6OmV4:-KJIfRp2awzUlewxJKJZ:-864019388) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/ConditionalPathPoint.cpp#L79)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> The connectee name specified here would ideally be the path to the coordinate, not just its name. Don't worry about it for this PR though.
</details>
True. But the update code cannot get the fullPathName() from old models, just its `coordinate` property value (string).

---

*[OpenSim/Simulation/Model/ConditionalPathPoint.h, line 76 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH8Nl2LKzEJiXschfH:-KJIgPP42EBt0chpixqk:1675637715) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/ConditionalPathPoint.h#L76)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> As far as I can tell, we do not have a convention of having a default value for `versionNumber`.
</details>
removed.

---

*[OpenSim/Simulation/Model/MovingPathPoint.cpp, line 337 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH9axFlkg7AKI3ycyr:-KJIdUnQur-5AN4n3wW7:-1954417350) ([raw file](https://github.com/opensim-org/opensim-core/blob/a0a0d99ab746cf9eae42d06509d74c56568af8e3/OpenSim/Simulation/Model/MovingPathPoint.cpp#L337)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Perhaps removing this check requires mention in the CHANGELOG.
</details>
It was not OK to remove. Thanks for flagging and making me double check. I restored and added more details in the comments.

---

*[OpenSim/Simulation/Model/MovingPathPoint.h, line 98 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHZJdoxReDGg2Ww1sq:-KJIc7zLBHbsVmChPut-:-499146009) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L98)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Also, the comment for `getLocation()` should be a doxygen comment.
</details>
Done.

---

*[OpenSim/Simulation/Model/Point.cpp, line 134 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHdA9oqzmtphXr9MKN:-KJIc4U-qsS9CgCX9tCE:-499146009) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/Point.cpp#L134)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Nitpicky: you could use `const auto r` and `const auto v`.
</details>
Done.

---

*[OpenSim/Simulation/Model/Point.h, line 111 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHbtk83p_tRa_2WxEU:-KJIbzGdME2QsO8QAMDb:-499146009) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/Point.h#L111)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> It would be clearer to me to say "the rate of change of the distance," if that is what it is. I'm fine with the name of the method being `calcSpeedBetween()` though. I'm not sure I understand the implementation of this method.
</details>
Done.

---

*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH7T6AYUkZrEGzafIq:-KJIc04pd1jSfzaA1NZg:-499146009) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Consider throwing an exception if the parent is not a GeometryPath.
</details>
Done.

---

*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHfvxPD3aI2agrTK7F:-KJIc2bGpfmsPfMtImGY:-499146009) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> Missing a call to Super::extendConnectToModel()?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJIexQTi6a1MvcVNc_Y:588181028)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
*[OpenSim/Simulation/Wrap/PathWrap.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJH7T6AYUkZrEGzafIq:-KJNTq1sbwExfcpj3I_8:215993149) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Wrap/PathWrap.cpp#L87)):*
<details><summary><i>Previously, aseth1 (Ajay Seth) wrote…</i></summary>
> Done.
</details>
No need to change this here, but this `msg` will be created even if the condition is false. In the future, I think we should avoid such overhead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, aseth1 (Ajay Seth) wrote…</i></summary>
> We should not have Path specific Points of any kind. They should be considered "path" Points by virtue of being subcomponents of a GeometryPath. Given the current pace of thinning out PathPoints and updates to GeometryPath this could be achieved before releasing 4.0.
> 
> I like you idea of overriding any user setting for location and making the location NaN until we have a proper fix in place.
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-:-KJNU2SqV_sJkNw24QFr:1417088721)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

chrisdembia commented:
I updated this PR so that the travis builds should pass.

I think @aseth1 is looking for one more reviewer for this PR.

aseth1 commented:
*[OpenSim/Simulation/Model/MovingPathPoint.h, line 112 \[r1\]](https://reviewable.io:443/reviews/opensim-org/opensim-core/990#-KJHj6BKGf0TjZINvZI6:-KKFJ9YREZTzFIkeQC0e:-556781233) ([raw file](https://github.com/opensim-org/opensim-core/blob/b942da9ff9db614004e6c6526989bdcf38a8ba80/OpenSim/Simulation/Model/MovingPathPoint.h#L112)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary>
> These methods should be `final` in Station, once we remove `Station` as `MovingPathPoint`'s base class.
</details>
Yes, I agree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opensim-org/opensim-core/990)*
<!-- Sent from Reviewable.io -->

aseth1 commented:
> I think @aseth1 is looking for one more reviewer for this PR.

Yes and looking to @aymanhab for feedback on the effect of these changes on the GUI.

aseth1 commented:
Thanks for merging in the updates @chrisdembia! Nice to see tests passing again. :clap:

aymanhab commented:
@aseth1 As long as the OpenSimContext class interface hasn't changed the GUI code will likely compile. I'm building against master to see the behavior before this PR and then will redo with this branch/PR.

aseth1 commented:
@aymanhab you are right that  I did not modify `OpenSimContext.h` but there were many changes to its implementation. I just wanted to see those changes exercised as a sanity check. If you feel `testContext` provides adequate coverage then are we OK to merge this PR?

aymanhab commented:
@aseth1 I'd say yes, please go ahead and merge. The closer we are to the final API the better we are in general, even if we need to go back and fix some implementation of methods on the OpenSimContext class.

aymanhab commented:
I'll open specific issues if I run into problems while testing.

aseth1 commented:
Thanks @aymanhab. 

@chrisdembia let me know if there is anything that needs a closer look. :microscope:

chrisdembia commented:
LGTM.

tkuchida commented on OpenSim/Simulation/Model/ConditionalPathPoint.cpp:
```diff
@@ -46,160 +46,90 @@ using namespace OpenSim;
- * underneath the model to match current version
+ * Set the coordinate that this point is linked to.
+ *
+ * @return Whether or not this point is active.
```
I don't think this `void()` should have a `@return` comment.

tkuchida commented on OpenSim/Simulation/Model/ConditionalPathPoint.cpp:
```diff
@@ -46,160 +46,90 @@ using namespace OpenSim;
  * @param aRange range min to change to.
  */
-void ConditionalPathPoint::setRangeMin(const SimTK::State& s, double aMin)
+void ConditionalPathPoint::setRangeMin(double minVal)
```
`@param aRange` should be changed to `@param minVal`

tkuchida commented on OpenSim/Simulation/Model/ConditionalPathPoint.cpp:
```diff
@@ -208,12 +138,9 @@ void ConditionalPathPoint::setRangeMin(const SimTK::State& s, double aMin)
  * @param aRange range max to change to.
  */
-void ConditionalPathPoint::setRangeMax( const SimTK::State& s, double aMax)
+void ConditionalPathPoint::setRangeMax(double maxVal)
```
`@param aRange` -> `@param maxVal` (or just delete the comment because it's obvious)

tkuchida commented on OpenSim/Simulation/Model/MovingPathPoint.cpp:
```diff
@@ -72,364 +63,211 @@ MovingPathPoint::~MovingPathPoint()
+    else // type == Constant
+        pInF[1] = get_y_location().calcValue(SimTK::Vector(1, 0.0));
 
     if (_zCoordinate) {
```
`_zCoordinate` -> `!_zCoordinate.empty()` to agree with changes to x and y cases?

tkuchida commented on OpenSim/Simulation/Model/MovingPathPoint.cpp:
```diff
@@ -47,472 +43,337 @@ using SimTK::Vec3;
+    return getParentFrame().getTransformInGround(s)*getLocation(s);
+}
+
+SimTK::Vec3 MovingPathPoint::calcVelocityInGround(const SimTK::State& s) const
```
Is there a test case for this method?

aseth1 replied:
I did not include additional unit tests. `testMomentArms` was my working test case for checking my implementation was correct (or at least produces consistent moment-arms).  These functions are also required to evaluate muscle path lengths and speeds (for models with MovingPathPoints) which are cases in `testForward`, `testStaticOptimization`, `testCMC`, etc...  

My intention is to eliminate `MovingPathPoint`, and have all Points provide their Jacobian with respect to any set of Coordinates. The implementation of `MovingPathPoint` makes little sense with exactly one Coordinate assigned to each component of the location vector. Once points can provide their partials I plan to test that more extensively.

tkuchida commented on OpenSim/Simulation/Model/MovingPathPoint.cpp:
```diff
@@ -47,472 +43,337 @@ using SimTK::Vec3;
+    // The velocity of the station in ground is a function of its frame's
+    // linear (vF = V_GF[1]) and angular (omegaF = V_GF[0]) velocity, such that
+    // velocity of the station: v = vF + omegaF x r
+    return V_GF[1] + V_GF[0] % r + v;
```
I'm confused here. Why isn't `r` the position vector from Ground to the origin of ParentFrame?

tkuchida replied:
Also, isn't this calculation missing the "omega cross r" contribution from the angular velocity of the point in the ParentFrame?

aseth1 replied:
> I'm confused here. Why isn't r the position vector from Ground to the origin of ParentFrame?

Because that isn't the local rotating vector due to the Frame's angular velocity. Think about it - if it was r to Ground origin you would get different velocity contribution by virtue of its location w.r.t Ground.  For example,  if we had two identical balls spinning at identical angular velocity you would get the velocity of a point on the surface to be different by virtue of where the origins of the spheres are w.r.t. ground - that cannot be right!

aseth1 replied:
>Also, isn't this calculation missing the "omega cross r" contribution from the angular velocity of the point in the ParentFrame?

Perhaps you overlooked line 367 and comments on lines 364-366?

tkuchida replied:
> Think about it

I did think about it. Based on your response, this method is meant to express the MovingPathPoint's local velocity in the Ground frame; in my opinion, such a method should not be named `calcVelocityInGround()`. I suggest (1) renaming the method to `expressLocalVelocityInGround()` or `calcLocalVelocityExpressedInGround()` and (2) including a description of the method on line 116 of MovingPathPoint.h.

aseth1 replied:
Bad choice of words. I meant imagine the scenario not that you did not think.

But you're still not right. Only `v` in the method is the local velocity expressed in ground. The value returned by the method is the velocity of the point in ground.

tkuchida replied:
> But you're still not right.

Yep, I was super confused. Thanks for clarifying.

tkuchida commented on OpenSim/Simulation/Model/MovingPathPoint.h:
```diff
@@ -101,49 +77,50 @@ OpenSim_DECLARE_CONCRETE_OBJECT(MovingPathPoint, PathPoint);
+    SimTK::Vec3 getLocation(const SimTK::State& s) const;
+    /** Get the local velocity of the MovingPathPoint w.r.t to and 
+        expressed in its Frame. To get the velocity of the point w.r.t.
+        and expressed in Ground, call getVelocityInGround(). */
```
I don't see a `getVelocityInGround()`. Was `calcVelocityInGround()` intended?

aseth1 replied:
> I don't see a getVelocityInGround(). Was calcVelocityInGround() intended?

`getVelocityInGround()` is implemented by the base and handles caching the calculated value. The calculation is virtual and the responsibility of the derived Point classes. Perhaps I missed an `override`? I will check.

tkuchida commented on OpenSim/Simulation/Model/PathPoint.cpp:
```diff
@@ -40,51 +40,25 @@ using namespace OpenSim;
+    upd_location() = currentFrame.findLocationInAnotherFrame(s, get_location(), body);
 
-    return newPoint;
+    // now assign this point's body to point to aBody
```
Perhaps clearer as `// now make "body" this PathPoint's parent Frame`

tkuchida commented:
`</review>`. I was confused by the implementation and comments in `MovingPathPoint::calcVelocityInGround()`; other suggestions are minor.

aseth1 commented:
@tkuchida thank you for the careful review! Please let me know how I might make the comments or implementation of `MovingPathPoint::calcVelocityInGround()` less confusing.

aseth1 commented:
@tkuchida I attempted to clarify the comments and steps for calculating the velocity of the moving point in Ground (`MovingPathPoint::calcVelocityInGround()`). Let me know if it is better now. Thanks!

tkuchida commented:
> Let me know if it is better now.

Ok, looks good :+1: Thanks for clarifying the comments.

aseth1 commented:
Woohoo!  Thank you @tkuchida.
